tips/tip-0003/tip-0003.md

```diff
@@ -0,0 +1,155 @@
+# TIP 0003: Subset Argument for Memory Consistency
+
+| TIP            | 0003                                               |
+|:---------------|:---------------------------------------------------|
+| authors:       | Alan Szepieniec                                    |
+| title:         | Subset Argument for Memory Consistency |
+| status:        | draft                                              |
+| created:       | 2022-08-25                                         |
+| issue tracker: |  |
+| pdf:           | [tip-0003.pdf](tip-0003.pdf)               |
+
+**Abstract.** In the current specification, the memory-like tables `RamTable`, `JumpStackTable`, and `OpStackTable` do not satisfy memory-consistency. Specifically, they are vulnerable to [Yuncong's attack](https://github.com/TritonVM/triton-vm/issues/12), which exploits the unverified and thus possibly-incorrect sorting in these tables. TIP-0002 addresses one part of the issue by introducing a new table argument, the *multi-table subset argument*. It establishes that certain values in different tables have a matching representative in a given lookup table. Applied to the present context, this technique establishes that every clock jump is positive.
+
+Unfortunately TIP-0002 introduces a lot of design complexity. The cause of this complexity is the Polynomial IOP for correct calculation of the formal derivative.
+
+This TIP proposes an alternative way to establish the requisite set relation, namely through a more traditional AIR/RAP construction that avoids computing the formal derivative entirely. While there is a performance penalty of two extra polynomial commitments, the end-result is a noticeably simpler design.
+
+TIPs 0002 and 0003 are mutually exclusive companions to TIP-0001, which introduces an new argument to establish the contiguity of regions of constant memory pointer. Together, TIP-0001 and (TIP-0002 or TIP-0003) fix the memory consistency issue.
+
+## Introduction
+
+How to establish that a clock jump is directed forward (as opposed to direct backward, which would indicate malicious behavior)? One strategy is to show that the *difference*, *i.e.*, the next clock cycle minus the current clock cycle, is itself a clock cycle. Recall that the Processor Table's clock cycles run from 0 to $T-1$. Therefore, valid differences belong to $\lbrace 1, ..., T-1 \rbrace$ and invalid ones to $\lbrace p - T + 1, ..., p - 1 \rbrace$.
```

Comment:
Is it worth pointing out that $T$ can never be so large as to be a problem for very far backwards jumps?